packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -100,14 +118,42 @@ if (!ExecutionEnvironment.canUseDOM) {
   let previousFrameTime = 33;
   let activeFrameTime = 33;
 
-  const frameDeadlineObject = {
+  // When a callback is scheduled, we register it by adding it's id to this
+  // object.
+  // If the user calls 'cIC' with the id of that callback, it will be
+  // unregistered by removing the id from this object.
+  // Then we skip calling any callback which is not registered.
+  // This means cancelling is an O(1) time complexity instead of O(n).
+  const registeredCallbackIds: Map<number, boolean> = new Map();
+
+  const frameDeadlineObject: Deadline = {
     didTimeout: false,
     timeRemaining() {
       const remaining = frameDeadline - now();
       return remaining > 0 ? remaining : 0;
     },
   };
 
+  const safelyCallScheduledCallback = function(callback, callbackId) {
+    if (!registeredCallbackIds.get(callbackId)) {
+      // ignore cancelled callbacks
+      return;
+    }
+    isCurrentlyRunningCallback = true;
+    try {
+      callback(frameDeadlineObject);
+      registeredCallbackIds.delete(callbackId);
+      isCurrentlyRunningCallback = false;
+    } catch (e) {
+      registeredCallbackIds.delete(callbackId);
+      isCurrentlyRunningCallback = false;
+      // Still throw it, but not in this frame.
+      setTimeout(() => {
```

Comment:
@sebmarkbage would like your input here, I think this is the last thing to decide before landing this PR. My previous comment here:

---
We discussed this in person and considered using postMessage to either throw the error sooner or even to create a postMessage event for each callback, so that errors could be thrown without interrupting the queue of callbacks getting called.

@sebmarkbage we would like you to make the call on whether the setTimeout is good enough for now, or should we explore using postMessage or some other approach?

I also considered saving the errors and just throwing them once we finish whatever is in the queue.

Reply:
Why not in this frame?

This will mess with debugging since it'll be a caught exception so you can't break point on it. Suspense will destroy the use of "pause on caught exception".

Another technique is to call this recursively with a try/finally and you don't need the catch.

Reply:
We talked a bit more and decided the 'try/finally' trick won't work, because a 2nd error thrown in the queue will get swallowed.

Going to look at alternatives, based on what we currently do in React to handle errors.

Reply:
![screen shot 2018-05-03 at 10 04 07 am](https://user-images.githubusercontent.com/1114467/39591706-6189e510-4eb9-11e8-8b3e-adb0b8f4d78c.png)
![screen shot 2018-05-03 at 10 03 59 am](https://user-images.githubusercontent.com/1114467/39591713-661965e2-4eb9-11e8-84c8-41714196231f.png)